File: js/384.js
/**
 * @param {number[]} nums
 */
var Solution = function (nums) {

  this._originNums = nums
};

/**
 * Resets the array to its original configuration and return it.
 * @return {number[]}
 */
Solution.prototype.reset = function () {

  return this._originNums
};

/**
 * Returns a random shuffling of the array.
 * @return {number[]}
 */
Solution.prototype.shuffle = function () {

  const nums = this._originNums.slice(0)
  const len = nums.length
  for (let i = 0; i < len; i++) {
    const j = Math.floor(Math.random() * len)
    const tmp = nums[i]
    nums[i] = nums[j]
    nums[j] = tmp
  }
  return nums
};

var _ = require('lodash');
var s = new Solution(_.range(0, 10000000))
s.shuffle()

File: py/384.py
import unittest
from random import randint

def shuffle(nums):
    
    size = len(nums)
    for i in range(size):
        j = randint(0, size - 1)
        [nums[i], nums[j]] = [nums[j], nums[i]]
    return nums

class Random:
    def __init__(self, nums):
        self.rands = shuffle(nums)
        self.i = 0
    
    def next(self):
        r = self.rands[self.i]
        self.i += 1
        return r

class Solution:

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.originNums = nums[:]
        self.nums = nums[:]

    def reset(self):
        """
        Resets the array to its original configuration and return it.
        :rtype: List[int]
        """
        self.nums[:] = self.originNums[:]
        return self.nums

    def shuffle(self):

        """
        Returns a random shuffling of the array.
        :rtype: List[int]
        """
        # jstart = randint(1, len(self.nums))
        # print("jstart:%d" % (jstart))
        for i in range(len(self.nums)):
            j = randint(0, len(self.nums) - 1)
            # print("i:%d, j:%d" % (i, j))
            [self.nums[i], self.nums[j]] = [self.nums[j], self.nums[i]]
        return self.nums



class SolutionTest(unittest.TestCase):
    def test(self):

        # s = Solution(list(range(100000)))
        # for i in range(10):
        #     s.shuffle()
        #     s.reset()
        s = Solution([-6,10,184])
        for i in range(100):
            print(s.shuffle())
            # print(s.reset())
            s.reset()



if __name__ == "__main__":
    unittest.main()
